### DISCUSSION TITLE
Migrating from Microsoft.Azure.Mobile.Server to Microsoft.AspNetCore.Datasync

### DISCUSSION THREAD

**Discussion by Brian-Con** [Azure/azure-mobile-apps#664] [Q&A]
Hi,

I'm migrating a Sync API from **Microsoft.Azure.Mobile.Server** to **Microsoft.AspNetCore.Datasync** and I'm not clear on how best to ensure the same **PushResult** gets returned to the Mobile App. Take the following example of inserting a record into a database which results in a Primary Key Violation.

Code snippet from the **Microsoft.Azure.Mobile.Server** project:

```c#
public virtual async Task<IHttpActionResult> PostItem(T item)
        {
            try
            {
                Common.LogInsert(Request, LogController, item);
                T current = await InsertAsync(item);
                return CreatedAtRoute("Tables", new { id = current.Id }, current);
            }
            catch (HttpResponseException rex)
            {
                string message = await rex.Response.Content.ReadAsStringAsync();
                Common.LogError(Request, LogController, Common.Actions.INSERTING.ToString(), message);
                throw rex;
            }
            catch (Exception ex)
            {
                Common.LogError(Request, LogController, Common.Actions.INSERTING.ToString(), ex.Message);
                throw ex;
            }
        }
``` 
In the above code, a Primary Key Violation would be caught by **catch (HttpResponseException rex)** and the **PushResult.Status** returned to the Mobile App would be **System.Net.HttpStatusCode.Conflict**.

To mimic the above behaviour in the **Microsoft.AspNetCore.Datasync** project I have the following code:

```c#
public override async Task<IActionResult> CreateAsync([Microsoft.AspNetCore.Mvc.FromBody] T item, CancellationToken token = default)
        {
            try
            {
                var InsertResult = await base.CreateAsync(item, token);
                return InsertResult;
            }           
            catch (Exception originalEx)
            {
                Exception? exception = originalEx;
                bool isConflict = false;

                while (exception != null)
                {
                    if (!string.IsNullOrWhiteSpace(exception.Message) && exception.Message.ToUpper().Contains("CONSTRAINT"))
                    {
                        isConflict = true;
                    }
                    IMECCommon.LogError(Request, LogController, IMECCommon.Actions.INSERTING.ToString(), "Ex = " + exception.Message);
                    exception = exception.InnerException;
                }

                if (isConflict)
                {
                    throw new ConflictException(item);
                }
                else
                {
                    throw;
                }
            }
        }
``` 

If I was to simply return the Exception caught by **catch(Exception)**, or even a **catch(RepositoryException)**, then the **PushResult.Status** returned to the Mobile App would be **System.Net.HttpStatusCode.InternalServerError**.

Thank you in advance.
Brian

**Comment by adrianhall** [comment]
What's the actual problem?  I can't discern the concern here.

Since you are overriding the CreateAsync(), it's up to you how to handle it.

**Reply by Brian-Con** [reply]
Hi Adrian,

Thanks for your reply.

The problem is that Result returned by **CreateAsync()** is not consistent between **Microsoft.Azure.Mobile.Server** and **Microsoft.AspNetCore.Datasync** and this inconsistency is breaking the logic in my Mobile App. My major concern is for the other inconsistences I've yet to discover. 

In the sample I provided, I'm having to do the extra step of scrutinizing the Result for the the text "CONSTRAINT" in order to determine if I should return a **ConflictException**, which is the **PushResult.Status** expected by my Mobile App when the error is due to a database constraint, such as a Primary Key Violation or Foreign Key Violation. For a **Conflict** my Mobile App performs a **CancelAndDiscardItemAsync()**. For a **InternalServerError** (which is what gets returned if I don't do the extra step) it will continuously re-try syncing the data.

Yes, I'm overriding the **CreateAsync()**, but ultimately I'm calling the **base.CreateAsync()** to do the work, and it is the Result from the call to the **base.CreateAsync()** that is not returning a Result consistent between **Microsoft.Azure.Mobile.Server** and **Microsoft.AspNetCore.Datasync**.

I posted my original question here seeking advice on how best to ensure the same **PushResult** gets returned to the Mobile App. I thought perhaps I'm doing something fundamentally wrong in how I'm doing things in **Microsoft.AspNetCore.Datasync** which could be highlighted to me.

Thanks for you time,
Brian

**Reply by adrianhall** [reply]
You may want to rethink the implementation there.  You can fork the EntityTableRepository (and EntityTableData) to create your own repository that throws the ConflictException when there is a primary key violation instead (or whatever other checks are required). By default, if the backend data source produces an error, we propagate that error through the stack which results in the Internal Server Error.  If you capture that at the database layer, then you can handle it explicitly.

In the older (4.2.0) days, the database layer was fairly opaque.  These days, you can easily write a new repository layer to do what you need.
